### ISSUE TITLE
Diagnosis "-Wuninitialized" fails when a variable is passed as a forwarding (universal) reference.

### ISSUE THREAD

**Issue by haruomaki** [llvm/llvm-project#63734]
When an uninitialized variable is passed to a function that takes a forwarding reference `T&&`, the warning message like "variable is not initialized" doesn't appear.

```C++
#include <iostream>

template <typename T>
void show(T &&x) {
    std::cout << std::forward<T>(x) << std::endl;
}

int main() {
    int x;
    show(x); // should emit warning, but doesn't
}
```

If you use `const T&` instead of `T&&`, you will see the warning *-Wuninitialized-const-reference* just fine as follows. 

```
assert.cpp:10:10: warning: variable 'x' is uninitialized when passed as a const reference argument here [-Wuninitialized-const-reference]
   10 |     show(x);
      |          ^
1 warning generated.
```

Please see also [Godbolt](https://godbolt.org/z/nM8xdn8Yz).

Is this behavior a bug, or is it intended?

**Comment by shafik** [comment]
I will note that gcc also does not diagnose this case: https://godbolt.org/z/vzb9obMxb

Note sure if this is intentional.

I think we know the analysis need some work in general: https://github.com/llvm/llvm-project/issues/49411
